warehouse/query-core/src/main/java/datawave/query/function/LimitFields.java

```diff
@@ -87,89 +88,127 @@ public Multimap<String,Attribute<? extends Comparable<?>>> load(String key) {
+                
+                // short circuit if we are not actually limiting this field.
+                // this is keeping with the original logic where a negative limit means to keep only hits
+                if (limit < 0) {
+                    continue;
+                }
+                
+                int missesToSet = Math.min(limit - get(keyNoGrouping, countKeepersForFieldMap), get(keyNoGrouping, countMissesRemainingForFieldMap));
+                
+                // if we have misses yet to keep
+                if (missesToSet > 0) {
+                    
+                    Attribute<?> attr = de.getValue();
+                    if (attr instanceof Attributes) {
+                        Attributes attrs = (Attributes) attr;
+                        Set<Attribute<? extends Comparable<?>>> attrSet = attrs.getAttributes();
+                        
+                        for (Attribute<? extends Comparable<?>> value : attrSet) {
+                            // if this was a miss previously set to not keep
+                            if (!value.isToKeep()) {
+                                value.setToKeep(true);
+                                increment(keyNoGrouping, countKeepersForFieldMap);
+                                decrement(keyNoGrouping, countMissesRemainingForFieldMap);
+                                missesToSet--;
+                                if (missesToSet == 0) {
+                                    break;
+                                }
+                            }
+                        }
+                    } else {
+                        // if this was a miss previously set to not keep
+                        if (!attr.isToKeep()) {
+                            attr.setToKeep(true);
+                            increment(keyNoGrouping, countKeepersForFieldMap);
+                            decrement(keyNoGrouping, countMissesRemainingForFieldMap);
+                            missesToSet--;
+                        }
+                    }
                 }
             }
-            if (log.isTraceEnabled()) {
-                log.trace("reducedMap:" + reducedMap);
-                log.trace("mapOfHits:" + hits.asMap());
-                log.trace("mapOfMisses:" + misses.asMap());
-            }
-            
+        }
+        
+        // reduce the document to those to keep
+        document.reduceToKeep();
+        
+        for (String keyNoGrouping : countForFieldMap.keySet()) {
             // only generate an original count if a field was reduced
-            if (countForFieldMap.get(keyNoGrouping) > this.limitFieldsMap.get(keyNoGrouping)) {
-                limitedFieldCounts.put(keyNoGrouping + ORIGINAL_COUNT_SUFFIX, countForFieldMap.get(keyNoGrouping));
+            if (get(keyNoGrouping, countForFieldMap) > get(keyNoGrouping, countKeepersForFieldMap)) {
```

Comment:
Slight readability improvement would be to assign these two values to variables. You could reuse the first one:

```
int origFieldCount = get(keyNoGrouping, countForFieldMap);
int keepFieldCount = get(keyNoGrouping, countKeepersForFieldMap);

if (origFieldCount > keepFieldCount) 
  // ... use origFieldCount
```

Reply:
good idea.  I will apply that suggestion throughout the class.